src/app.ts

```diff
@@ -176,16 +173,94 @@ export async function newApp(
   // Setup routes
   //
 
-  app.get("/login", spidAuth);
-
+  await registerLoginRoute(app);
   registerPublicRoutes(app);
   registerAuthenticationRoutes(app, authenticationBasePath);
   registerAPIRoutes(app, APIBasePath, allowNotifyIPSourceRange);
   registerPagoPARoutes(app, PagoPABasePath, allowPagoPAIPSourceRange);
 
+  const refreshTimeMilliseconds =
+    container.resolve<number>(IDP_REFRESH_INTERVAL_SECONDS) * 1000;
+  const idpMetadataUpdateTimer = idpMetadataUpdater(
+    app,
+    refreshTimeMilliseconds
+  );
+  app.on("server:stop", () => {
+    clearInterval(idpMetadataUpdateTimer);
+  });
   return app;
 }
 
+/**
+ * Initializes SpidStrategy for passport and setup /login route.
+ */
+async function registerLoginRoute(app: Express): Promise<void> {
+  // Add the strategy to authenticate the proxy to SPID.
+  const spidStrategy = await container.resolve<Promise<passport.Strategy>>(
+    SPID_STRATEGY
+  );
+  passport.use("spid", spidStrategy);
+  const spidAuth = passport.authenticate("spid", { session: false });
+  app.get("/login", spidAuth);
+}
+
+/**
+ * Clears SPID_STRATEGY cache from awilix.
+ * Then /login route will be overridden with an updated SPID_STRATEGY.
+ */
+async function clearAndReloadSpidStrategy(
+  app: Express,
+  onRefresh?: () => void
+): Promise<void> {
+  log.info("Started Spid strategy re-initialization ...");
+
+  // Clear cache for SPID_STRATEGY definition (singleton lifetime) on awilix.
+  // The next resolution of SPID_STRATEGY will reinstantiate the registered class.
+  container.cache.delete(SPID_STRATEGY);
+
+  passport.unuse("spid");
+
+  // tslint:disable-next-line: no-any
+  const isLoginRoute = (route: any) => {
+    if (
+      route.route &&
+      route.route.path === "/login" &&
+      route.route.methods &&
+      route.route.methods.get
+    ) {
+      return true;
+    }
+    return false;
+  };
+
+  // Remove /login route from Express router stack
+  // tslint:disable-next-line: no-object-mutation
+  app._router.stack = app._router.stack.filter(not(isLoginRoute));
+
+  await registerLoginRoute(app);
+  log.info("Spid strategy re-initialization complete.");
+  if (onRefresh) {
+    onRefresh();
+  }
+}
+
+/**
+ * Sets an interval to reload SpidStrategy
+ */
+export function idpMetadataUpdater(
```

Comment:
let's make it explicit that we're starting something that will continue after the function ends:

```suggestion
export function startIdpMetadataUpdater(
```

Reply:
Fixed in a6e4c2a